### ISSUE TITLE
Add option to cy.route2() to match request body

### ISSUE THREAD

**Issue by asumaran** [cypress-io/cypress#8560]
I was trying to `cy.wait()` some specific GraphQL requests that were made using `fetch` and using `cy.route2()` but I don't see an option to add a matcher for the request body. [According to the type definition there's not `body` param](https://github.com/cypress-io/cypress/blob/3f901fcb7084d708eecee7116863c15d161c8ae6/packages/net-stubbing/lib/external-types.ts#L126-L167
)

The closest thing is `query` but it doesn't consider the request body. Is this by design or there are plans to add it?

### Current behavior:

All GraphQL queries go to the same endpoint "server.com/graphql" using the `POST` method. The query itself is sent in the request payload and it's a JSON-like string as follows:

```gql
{
  "query": "query ($limit: Int) {
    searchShippingTemplate(limit: $limit) {
      shippingTemplates {
        id
      }
    }
  }",
  "variables": {
    "limit": 25
  }
}
```

My plan was add a match for  `searchShippingTemplate` like:

```js
cy.route2({
  url: "http://localhost:3000/graphql",
  method: "POST",
  body: "*searchShippingTemplate*",
}).as("getQuery");
```

But there's no way to do that.

### Desired behavior:

Add `body` option to `RouteMatcherOptions` in order to match a string in the request body.

It would be even better to be able to pass a function (for debugging purposes) since there's no easy way to know what exactly is `.route2()` making the comparison against for each `RouteMatcherOptions` option. 

```js
cy.route2({
  url: "http://localhost:3000/graphql",
  method: "POST",
  body: (bodyString) => {
    console.log(bodyString); // In theory we could even mutate the request
    return bodyString;
  },
}).as("getQuery");

```

### Test code to reproduce

This is not a bug and more like a feature request. 

### Versions

```
"cypress": "^5.1.0"
```

**Comment by flotwig** [comment]
@asumaran yes, this would be nice to add, it is currently not implemented but as you say it would help with some implementations.

You can modify the request body and response body already.

```js
// modify an outgoing request
cy.route2({
  url: "http://localhost:3000/graphql",
  method: "POST"
}, (req) => {
  console.log(req.body)
  req.body.replace('foo, 'bar') // modifies outgoing request
})
```

You can also use a deferred promise to `.wait` on a request that you've "matched" via the request handler.

```js
// wait on a dynamically-matched request using a deferred Promise
const p = Cypress.Promise.defer()

cy.route2({
  url: "http://localhost:3000/graphql",
  method: "POST"
}, (req) => {
  if (req.body.includes('searchShippingTemplate')) {
    p.resolve() // resolve the deferred promise
  }
})
// ... do some more Cypress stuff here ...
// now, wait on that deferred Promise to resolve:
cy.wrap(p)
```

I hope this helps in the meantime while we add a more formal API around this. :)

**Comment by asumaran** [comment]
Thank you @flotwig. I’m going to try your suggestions.

**Comment by AnderssonChristian** [comment]
I'm interested in this as well.

**Comment by jandrade** [comment]
+1 It would be great to add this feature for the same reasons commented here.

**Comment by vrknetha** [comment]
It would be a great feature to match GraphQl queries.

**Comment by m4dc4p** [comment]
@flotwig given `Promise.defer()` is deprecated, is there any way to do this with aliases? Does returning a promise that is resolved/rejected based on the request work?

```ts
cy.route2({
  url: "http://localhost:3000/graphql",
  method: "POST"
}, (req) => {
  return new Promise((resolve, reject) => {
    if (req.body.includes('searchShippingTemplate')) {
      resolve() // resolve the deferred promise
    } else {
      reject()
    }
})}).as(`searchShipping`)

cy.wait(`@searchShipping`)
```

**Comment by hartzis** [comment]
📓 @bahmutov wrote a great post about graphQL requests and `cy.route2` and matching request body's.
- https://glebbahmutov.com/blog/smart-graphql-stubbing/

📓 @m4dc4p Additionally there apparently is a new way to alias specific requests now too.
- https://docs.cypress.io/api/commands/route2.html#Aliasing-individual-requests
```
cy.route2('POST', '/graphql', (req) => {
  if (req.body.includes('mutation')) {
    req.alias = 'gqlMutation'
  }
})
```

**Comment by m4dc4p** [comment]
> 📓 @bahmutov wrote a great post about graphQL requests and `cy.route2` and matching request body's.
> 
>     * https://glebbahmutov.com/blog/smart-graphql-stubbing/

I've read it. Unfortunately, it does not cover matching specific requests ...

> 📓 @m4dc4p Additionally there apparently is a new way to alias specific requests now too.
> 
>     * https://docs.cypress.io/api/commands/route2.html#Aliasing-individual-requests

🤯 - that looks like what I need. Thank you!

**Comment by flotwig** [comment]
> @flotwig given `Promise.defer()` is deprecated, is there any way to do this with aliases? Does returning a promise that is resolved/rejected based on the request work?

@m4dc4p bah, I wish they hadn't stuck that deprecated warning on `defer`, `defer` is totally fine to use in situations like this imo.

You can create your own deferred promise:

```
function deferredPromise() {
	let resolve, reject
	const promise = new Cypress.Promise((_resolve, reject) => {
		resolve = _resolve
		reject = _reject
	})
	return { resolve, reject, promise }
}
```

Works the same as `Promise.defer`.

https://docs.cypress.io/api/commands/route2.html#Aliasing-individual-requests also works well, especially if you need to match on dynamic criteria.

**Comment by mehrad77** [comment]
This link is dead:  https://docs.cypress.io/api/commands/route2.html#Aliasing-individual-requests

**Comment by jennifer-shehane** [comment]
@mehrad77 https://on.cypress.io/intercept#Aliasing-individual-requests

**Comment by antonlegkiy** [comment]
+1
I have a similar situation when POST requests query params are the same, but actions are located in the payload, so it would be nice to catch only requests with action that you're waiting for.

**Comment by aturkewi** [comment]
+1 here as well. This would be great to have.

**Comment by ChapDDR** [comment]
Another +1

**Comment by locinus** [comment]
+1

**Comment by phattanun** [comment]
+1

**Comment by ozthekoder** [comment]
This is really needed
+1

**Comment by andi-dev** [comment]
Maybe the easiest way to support this and potentially many other use cases, would be if `intercept` would optionally take a function as the first parameter, and expect this function to return a boolean. (true = request matched, false = request did not match).

I am thinking of something like this:

```
  cy.intercept((req) => {
    req.method === 'POST' && req.url.match('some-url-fragment') && req.body.match('someParam')
  })
```

**Comment by bahmutov** [comment]
I think it would be nice to do what the request header matching does - you just need one of the headers to match. I often use it to stub GraphQL resources using the custom "X-operation-name" header

from https://github.com/bahmutov/todo-graphql-example/blob/master/cypress/integration/intercept-spec.js
```
    // we have special middleware in our GraphQL client
    // that puts the operation name in the request header "x-gql-operation-name"
    // we can define intercepts using this custom header
    cy.intercept({
      method: 'POST',
      url: '/',
      headers: {
        'x-gql-operation-name': 'allTodos',
      },
    }).as('allTodos')

    cy.intercept({
      method: 'POST',
      url: '/',
      headers: {
        'x-gql-operation-name': 'AddTodo',
      },
    }).as('addTodo')

    cy.intercept({
      method: 'POST',
      url: '/',
      headers: {
        'x-gql-operation-name': 'updateTodo',
      },
    }).as('updateTodo')
```

We cannot modify the application code like this, so I would love to be able to match by a part of the request body. In GraphQL requests I have `operationName` field for example and I would love to use it. Maybe something like `bodyPart` which could be a nested object; if the request body includes it, then the matcher fires?

```js
 cy.intercept({
    method: 'POST',
    url: '/',
    bodyPart: {
        'operationName': 'allTodos',
    },
}).as('allTodos')
```

**Comment by johnmiroki** [comment]
> Maybe the easiest way to support this and potentially many other use cases, would be if `intercept` would optionally take a function as the first parameter, and expect this function to return a boolean. (true = request matched, false = request did not match).
> 
> I am thinking of something like this:
> 
> ```
>   cy.intercept((req) => {
>     req.method === 'POST' && req.url.match('some-url-fragment') && req.body.match('someParam')
>   })
> ```

This would be the simplest yet most versatile solution. We have $batch request for odata where the endpoint stays the same while the request path is specified in the request body (or request body part if you will). If the function matcher were to be supported, we could write our own requestMatcher with ease.

**Comment by tripflex** [comment]
+1

**Comment by mcollins2-quantium** [comment]
+1

**Comment by Grutula** [comment]
This is really needed
+1

**Comment by amyzhao-seismic** [comment]
+1

**Comment by Baudry-G** [comment]
This is really needed
+1

**Comment by ohirnyak** [comment]
+1

**Comment by lokriet** [comment]
+1

**Comment by ilibilibom** [comment]
+1

**Comment by shinsid** [comment]
+1

**Comment by coolduebtn** [comment]
+1 . I am surprised that this is still open and the team has not implemented this yet

**Comment by aturkewi** [comment]
+1. Please, this would be really helpful to have.

**Comment by arm22** [comment]
+1

**Comment by DavitMkhitaryan** [comment]
+1

**Comment by ercgrat** [comment]
+1

**Comment by hamidjafari** [comment]
+1

**Comment by agrandin0** [comment]
+1

**Comment by aowss** [comment]
+1

**Comment by simplyaboutcode** [comment]
+1

**Comment by GraceZaborski** [comment]
+1

**Comment by ahlivehr** [comment]
+1

**Comment by dlangsam** [comment]
+1

**Comment by rahilarparkar** [comment]
+1

**Comment by lawrensylvan** [comment]
+1
This is really needed, i can't believe I have to intercept all calls of the same url with the same handler if their request body differ. I have to catch them all and then return a different response body depending on the received request body, all in the same function. I would rather be able to write separate intercept for that.

**Comment by Nastya9js** [comment]
+1
